### DISCUSSION TITLE
Best Practices for Efficiently Importing Large Volumes of Data into NetBox

### DISCUSSION THREAD

**Discussion by JessicaRigol** [netbox-community/netbox#16443] [Q&A]
I am starting to use NetBox in my company and need to fill in a large volume of data in a detailed and efficient manner. What best practices or tools do you recommend to facilitate this process?

**Comment by PaulR282** [comment]
This really depends on the data source you are using. I usually throw together a python script and use the RestAPI to import the data.

**Reply by masterivanic** [reply]
This approach take more time to ingest netbox data. I see that netbox release a solution call TurboBulk to achieve this as well. But this solution is not open source

**Comment by svalouch** [comment]
I was facing the same problem with thousands of devices and their numerous components, and ended up creating a custom django management command: https://docs.djangoproject.com/en/6.0/howto/custom-management-commands/
To do that, I've created a minimal plugin to house the command, which I then call via `python3 /path/to/netbox/manage.py mycustomcommand <inputfile>`. The command is not reachable via API or UI, so no issues with keeping it secure.

The folder structure looks like that:
```
pyproject.toml
myplugin/
  __init__.py  # just the bare minimum plugin config
  management/
    commands/
      mycustomcommand.py
```

Inside the command, you have the entire functionality of Netbox and Django at your disposal, without having to go through the API. Since there are no models created, there is no need to do migrations and you can safely remove the plugin once you're done.

Things worthy of note when going with a management command:

* You bypass all permissions and at least some of the validation, so be very careful.
* Do not use Django's `get_or_create`-function (especially if you expect to redo the import to update data, even if it looks like the perfect candidate). Using it bypasses important Netbox functions and things can break further down the line. Instead, check if an object exists in the DB first, and if not then create it explicitly:
  ```python
  newdev = Device(name=…, …)
  newdev.full_clean()
  newdev.save()
  ```
* No changelog entries will be created for you, you'll have to create them yourself. I found that creating a single, fresh UUID at the start of the command and setting a static username to the name of the script is giving me the best results later: Netbox groups all changelog entries under the UUID, so I can easily see everything the script touched in an invocation, and the fabricated username makes it clear that it was a custom script (and which one) rather than through API or UI.
